<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Zining Qi, Alexandre Pelletier" />


<title>Factor Interpretability Analysis on gEBMF</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Missing Data Imputation for Multi-Omics Data</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/zq2209/omics-imputation-paper">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Factor Interpretability Analysis on
gEBMF</h1>
<h4 class="author">Zining Qi, Alexandre Pelletier</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-04-14
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong>
<code>omics-imputation-paper/analysis/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed1code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomzq2209omicsimputationpapertree5cfa26a568d7bcae186a1d7a43000a6037ed5f3btargetblank5cfa26aa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/zq2209/omics-imputation-paper/tree/5cfa26a568d7bcae186a1d7a43000a6037ed5f3b" target="_blank">5cfa26a</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomzq2209omicsimputationpapertree5cfa26a568d7bcae186a1d7a43000a6037ed5f3btargetblank5cfa26aa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/zq2209/omics-imputation-paper/tree/5cfa26a568d7bcae186a1d7a43000a6037ed5f3b" target="_blank">5cfa26a</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown
(<code>analysis/factor_interpretability.Rmd</code>) and HTML
(<code>docs/factor_interpretability.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/zq2209/omics-imputation-paper/blob/5cfa26a568d7bcae186a1d7a43000a6037ed5f3b/analysis/factor_interpretability.Rmd" target="_blank">5cfa26a</a>
</td>
<td>
“John
</td>
<td>
2025-04-14
</td>
<td>
update
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/zq2209/omics-imputation-paper/e8f39b394d7753b7565c7a69f40422d9da1620c4/docs/factor_interpretability.html" target="_blank">e8f39b3</a>
</td>
<td>
“John
</td>
<td>
2025-04-13
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/zq2209/omics-imputation-paper/blob/f911568148a684bc7b865d05c1efb26468fa9c38/analysis/factor_interpretability.Rmd" target="_blank">f911568</a>
</td>
<td>
“John
</td>
<td>
2025-04-13
</td>
<td>
update
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>This analysis aims to evaluate and compare two dimensionality
reduction methods - generalized Empirical Bayes Matrix Factorization
(gEBMF) and Principal Component Analysis (PCA) - for interpreting omics
data (i.e. methylation). The primary objective is to determine which
method better captures known biological and technical variation in the
data. By systematically assessing how factors from each method associate
with clinical variables (age, sex, AD diagnosis) and technical
covariates (batch effects, processing variables), we can make informed
decisions about which dimensionality reduction approach is most suitable
for downstream analyses. The results will help us understand whether the
more sophisticated gEBMF method provides meaningful advantages over
traditional PCA for omics data interpretation.</p>
<div id="setup-and-initialization" class="section level1">
<h1>Setup and Initialization</h1>
<pre class="r"><code># Create output directory for results
out &lt;- &#39;methylation/outputs/08-flashier_factors_interpretability&#39;
dir.create(out, recursive = TRUE, showWarnings = FALSE)

# Load required packages
library(tidyverse)
library(data.table)
library(flashier)
library(knitr)

# Source utility functions
source(&#39;code/r_utils.R&#39;)
source(&#39;code/pca_utils.R&#39;)</code></pre>
<div id="data-preparation" class="section level3">
<h3>Data Preparation</h3>
<div id="load-and-process-metadata" class="section level5">
<h5>Load and Process Metadata</h5>
<p>We start by reading three key metadata files: clinical data
containing participant characteristics and diagnoses, biospecimen
information about sample collection and processing, and
methylation-specific technical metadata. These datasets are merged using
specimen identifiers to create a comprehensive metadata table and
filtered to include only samples with valid study information.</p>
<pre class="r"><code>mtd_clin &lt;- fread(&#39;../ref-data/ROSMAP/ROSMAP_clinical.csv&#39;)
mtd_bio &lt;- fread(&#39;../ref-data/ROSMAP/ROSMAP_biospecimen_metadata.csv&#39;)
mtd_meth &lt;- fread(&#39;../ref-data/ROSMAP/ROSMAP_assay_methylationArray_metadata.csv&#39;)

mtd &lt;- merge(mtd_clin, 
             mtd_bio[assay == &#39;methylationArray&#39; &amp; individualID != &#39;&#39;][, .(individualID, specimenID)])
mtd &lt;- merge(mtd, 
             mtd_meth[, .(specimenID, batch, Sentrix_ID, Sentrix_Row_Column)], 
             by = &#39;specimenID&#39;)

mtdf &lt;- mtd[Study != &#39;&#39;]</code></pre>
</div>
<div id="process-clinical-variables" class="section level5">
<h5>Process Clinical Variables</h5>
<p>The clinical data processing step transforms raw variables into
analysis-ready formats:</p>
<ul>
<li><p>Age at death is converted to numeric values, with special
handling for the “90+” category which is coded as 91.</p></li>
<li><p>Study identifiers are converted to numeric factors for easier
modeling.</p></li>
<li><p>Sample IDs are standardized to ensure compatibility with R
syntax.</p></li>
<li><p>For Alzheimer’s disease (AD) analysis:</p>
<ul>
<li><p>Raw cogdx scores, a binary dementia indicator</p></li>
<li><p>three-category AD status variable.</p></li>
<li><p>APOE genotype, a major genetic risk factor for AD</p></li>
<li><p>Race</p></li>
<li><p>APOE4 carrier status variable.</p></li>
</ul></li>
</ul>
<pre class="r"><code>mtdf[, age_death_num := as.numeric(ifelse(age_death == &#39;90+&#39;, 91, age_death))]
mtdf[, Study.num := as.numeric(factor(Study))]
mtdf[, specimenID_corr := make.names(specimenID)]
mtdf[, AD_status := cogdx]
mtdf[, dementia := !cogdx %in% 1:3]
mtdf[, apoe4_career := apoe_genotype %in% c(44, 34, 24)]
mtdf[, apoe_genotype := factor(apoe_genotype, levels = c(22, 23, 24, 33, 34, 44))]
mtdf[, apoe_genotype_num := as.numeric(apoe_genotype)]
mtdf[, race_bin := race == 1]
mtdf[, AD_status := sapply(cogdx, function(x) {
  ifelse(x == 1, 1, ifelse(x %in% 2:3, 2, ifelse(x %in% 4:5, 3, NA)))
}]</code></pre>
</div>
</div>
</div>
<div id="factor-analysis" class="section level1">
<h1>Factor Analysis</h1>
<div id="load-gebmf-factors" class="section level2">
<h2>Load gEBMF Factors</h2>
<p>The factor loadings (matrix F) from gEBMF analysis are extracted,
which represent the weights of each factor across samples. We focus on
the first 50 factors, which typically capture the majority of meaningful
variation.</p>
<pre class="r"><code>fl &lt;- readRDS(&#39;../fungen-xqtl-analysis/analysis/TCW_BU/ROSMAP/mqtl/outputs/01-ROSMAP_methylation_data_preprocessing/flashier_opti_factors.rds&#39;)
plot(fl, main = &quot;gEBMF Factor Analysis Results&quot;)
ldf &lt;- ldf(fl, type = &#39;f&#39;)
fs_dt &lt;- data.table(ldf$F[, 1:50], keep.rownames = &#39;specimenID_corr&#39;)
setnames(fs_dt, old = paste0(&#39;V&#39;, 1:50), new = paste0(&#39;f&#39;, 1:50))
fs_dt_melt &lt;- melt(fs_dt, measure.vars = paste0(&#39;f&#39;, 1:50), 
                   variable.name = &#39;factor&#39;, value.name = &#39;factor.value&#39;)
fs_dt_melt &lt;- merge(fs_dt_melt, data.table(factor = paste0(&#39;f&#39;, 1:50), ve = ldf$D[1:50]))</code></pre>
<div id="principal-component-analysis" class="section level3">
<h3>Principal Component Analysis</h3>
<div id="load-and-prepare-omics-data" class="section level5">
<h5>Load and Prepare Omics Data</h5>
<p>The omics data preparation involves several quality control steps,
methylation as an example:</p>
<ul>
<li><p>We first load the raw methylation beta values, which represent
the proportion of methylated alleles at each CpG site.</p></li>
<li><p>To ensure data quality, we remove CpG sites with excessive
missing values (&gt;50% missing across samples).</p></li>
<li><p>We exclude all probes on chromosome Y to avoid sex-specific
artifacts.</p></li>
<li><p>The remaining beta values are converted to a matrix format and
transformed using the probit function (qnorm), which approximates a
normal distribution better suited for linear analyses.</p></li>
</ul>
<pre class="r"><code>pheno &lt;- read_delim(&#39;../fungen-xqtl-analysis/analysis/TCW_BU/ROSMAP/mqtl/outputs/01-ROSMAP_methylation_data_preprocessing/ROSMAP_assay_methylationArray_metadata.sesame.beta.bed.gz&#39;, 
                   delim = &#39;\t&#39;)
pheno &lt;- pheno[rowMeans(is.na(pheno[, 5:ncol(pheno)])) &lt;= 0.5, ]
pheno &lt;- pheno[pheno$`#chr` != &#39;chrY&#39;, ]
x &lt;- as.matrix(data.frame(pheno[, 4:ncol(pheno)], row.names = &#39;ID&#39;))
x &lt;- qnorm(x)</code></pre>
</div>
<div id="perform-pca" class="section level5">
<h5>Perform PCA</h5>
<p>Principal Component Analysis is performed using a custom function
that implements soft imputation for missing values. The analysis
generates principal components that capture orthogonal directions of
maximum variance in the methylation data. We extract the first 50 PCs,
which typically account for the majority of biologically relevant
variation.</p>
<pre class="r"><code>pca &lt;- RunPca(x, imputation = &#39;soft&#39;)
pcs_dt &lt;- data.table(pca$x[, 1:50], keep.rownames = &#39;specimenID_corr&#39;)
pcs_dt_melt &lt;- melt(pcs_dt, measure.vars = paste0(&#39;PC&#39;, 1:50), 
                   variable.name = &#39;factor&#39;, value.name = &#39;factor.value&#39;)
pcs_dt_melt &lt;- merge(pcs_dt_melt, data.table(factor = paste0(&#39;PC&#39;, 1:50), 
                     ve = pca$sdev[1:50]^2)</code></pre>
</div>
</div>
<div id="combined-analysis" class="section level3">
<h3>Combined Analysis</h3>
<div id="merge-data-sources" class="section level5">
<h5>Merge Data Sources</h5>
<p>We stack the gEBMF factors and PCA components, adding a method
identifier to distinguish their origins. This combined dataset is then
joined with the processed clinical metadata, creating a comprehensive
table linking each factor/PC score to all relevant covariates.</p>
<pre class="r"><code>factors &lt;- rbind(
  fs_dt_melt[, method := &#39;gEBMF&#39;],
  pcs_dt_melt[, method := &#39;PCA&#39;]
)
factors &lt;- merge(factors, mtdf, by = &#39;specimenID_corr&#39;)
factors &lt;- melt(factors, 
               measure.vars = c(&#39;age_death_num&#39;, &#39;Study.num&#39;, &#39;dementia&#39;, &#39;AD_status&#39;,
                               &#39;apoe4_career&#39;, &#39;apoe_genotype_num&#39;, &#39;race_bin&#39;,
                               &#39;spanish&#39;, &#39;braaksc&#39;, &#39;pmi&#39;, &#39;batch&#39;, &#39;msex&#39;),
               variable.name = &#39;known.cov&#39;, value.name = &#39;cov.value&#39;)
factors &lt;- factors[!is.na(cov.value)]
factors[, cov.value := as.numeric(cov.value)]</code></pre>
</div>
<div id="statistical-analysis" class="section level5">
<h5>Statistical Analysis</h5>
<p>The statistical analysis systematically evaluates associations
between factors/PCs and covariates. For each combination, we fit a
linear model and extract multiple metrics: R-squared (proportion of
variance explained), regression coefficients (effect sizes), standard
errors, p-values, and derived z-scores. These metrics provide
complementary perspectives on the strength and significance of each
association. We identify the top 10 most significant factors for each
covariate-method combination, and extract factor numbers for
visualization purposes.</p>
<pre class="r"><code>factors[, r2 := summary(lm(factor.value ~ cov.value))$adj.r.squared,
        by = .(factor, known.cov, method)]
factors[, beta := summary(lm(factor.value ~ cov.value))$coefficients[2, 1],
        by = .(factor, known.cov, method)]
factors[, beta_se := summary(lm(factor.value ~ cov.value))$coefficients[2, 2],
        by = .(factor, known.cov, method)]
factors[, pvalue := summary(lm(factor.value ~ cov.value))$coefficients[2, 4],
        by = .(factor, known.cov, method)]
factors[, z_score := beta / beta_se]

res &lt;- unique(factors[, .(factor, ve, known.cov, pvalue, r2, beta, beta_se, z_score, method)])
res[, factor := factor(factor, levels = c(paste0(&#39;f&#39;, 1:50), paste0(&#39;PC&#39;, 1:50)))]
res[, padj := p.adjust(pvalue, method = &#39;bonferroni&#39;), by = .(known.cov, method)]
res[, top := rank(pvalue) &lt;= 10, by = c(&#39;known.cov&#39;, &#39;method&#39;)]
res[, factor_num := str_extract(factor, &#39;[0-9]+&#39;) |&gt; as.numeric() |&gt; as.factor()]</code></pre>
</div>
</div>
<div id="visualization" class="section level3">
<h3>Visualization</h3>
<div id="top-factors-by-covariate" class="section level5">
<h5>Top Factors by Covariate</h5>
<p>The visualization step creates a comprehensive plot comparing how
factors from each method associate with key covariates. We focus on six
important variables representing both technical (pmi, batch) and
biological (AD_status, age, sex) sources of variation. The plot displays
absolute z-scores for the top 10 most significant factors/PCs for each
covariate, allowing direct comparison between gEBMF and PCA. A reference
line at z=3 indicates a rough significance threshold.</p>
<pre class="r"><code>covs_to_shows &lt;- c(&#39;pmi&#39;, &#39;batch&#39;, &#39;AD_status&#39;, &#39;apoe4_career&#39;, &#39;age_death_num&#39;, &#39;msex&#39;)

ggplot(res[top &amp; known.cov %in% setdiff(covs_to_shows, &#39;apoe4_career&#39;)][
         , known.cov := factor(known.cov, levels = covs_to_shows)]) +
  geom_col(aes(x = method, y = abs(z_score), fill = factor_num), position = &#39;dodge&#39;) +
  theme_bw() + 
  geom_hline(yintercept = 3, linetype = &#39;dashed&#39;, col = &#39;red3&#39;, linewidth = 0.5) +
  facet_wrap(&#39;known.cov&#39;, scales = &#39;free&#39;) +
  labs(title = &#39;Top 10 factors associated with known covariates&#39;,
       x = &#39;Dimensionality Reduction Method&#39;,
       y = &#39;Absolute Z-score&#39;,
       fill = &#39;Factor Number&#39;) +
  theme(legend.position = &#39;bottom&#39;)

ggsave(file.path(out, &#39;factors_known_cov_assoc_ROSMAP.pdf&#39;), width = 8, height = 6)</code></pre>
</div>
</div>
<div id="results-summary" class="section level3">
<h3>Results Summary</h3>
<p>The results summary provides quantitative comparisons between
methods. For each covariate, we calculate four key metrics: mean and
maximum z-scores (showing overall and peak association strength), mean
R-squared (variance explained), and number of significantly associated
factors. These metrics collectively indicate how well each
dimensionality reduction method captures the various sources of
variation.</p>
<pre class="r"><code>summary_table &lt;- res[top == TRUE, .(
  Mean_Z = mean(abs(z_score)),
  Max_Z = max(abs(z_score)),
  Mean_R2 = mean(r2),
  N_Sig = sum(padj &lt; 0.05)
), by = .(method, known.cov)]

kable(summary_table, caption = &quot;Summary of Factor-Covariate Associations&quot;)</code></pre>
<p>The summary table enables quick identification of which method
performs better for different covariate types (technical vs. biological)
and which covariates are most strongly reflected in the
reduced-dimensional representations.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.3 (2025-02-28)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.4

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] vctrs_0.6.5       cli_3.6.4         knitr_1.50        rlang_1.1.5      
 [5] xfun_0.51         stringi_1.8.7     promises_1.3.2    jsonlite_2.0.0   
 [9] workflowr_1.7.1   glue_1.8.0        rprojroot_2.0.4   git2r_0.36.2     
[13] htmltools_0.5.8.1 httpuv_1.6.15     sass_0.4.9        rmarkdown_2.29   
[17] evaluate_1.0.3    jquerylib_0.1.4   tibble_3.2.1      fastmap_1.2.0    
[21] yaml_2.3.10       lifecycle_1.0.4   whisker_0.4.1     stringr_1.5.1    
[25] compiler_4.4.3    fs_1.6.5          Rcpp_1.0.14       pkgconfig_2.0.3  
[29] rstudioapi_0.17.1 later_1.4.1       digest_0.6.37     R6_2.6.1         
[33] pillar_1.10.1     magrittr_2.0.3    bslib_0.9.0       tools_4.4.3      
[37] cachem_1.1.0     </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
