<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Zining Qi" />


<title>xQTL Analysis on Real Data</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Missing Data Imputation for Multi-Omics Data</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/zq2209/omics-imputation-paper">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">xQTL Analysis on Real Data</h1>
<h4 class="author">Zining Qi</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-04-13
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong>
<code>omics-imputation-paper/analysis/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity itâ€™s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed1code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomzq2209omicsimputationpapertreee7660fed228e6ac49956707505367ccb4e67dc57targetblanke7660fea">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/zq2209/omics-imputation-paper/tree/e7660fed228e6ac49956707505367ccb4e67dc57" target="_blank">e7660fe</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomzq2209omicsimputationpapertreee7660fed228e6ac49956707505367ccb4e67dc57targetblanke7660fea"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/zq2209/omics-imputation-paper/tree/e7660fed228e6ac49956707505367ccb4e67dc57" target="_blank">e7660fe</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/

</code></pre>
<p>
Note that any generated files, e.g.Â HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/real_data_xqtl.Rmd</code>) and
HTML (<code>docs/real_data_xqtl.html</code>) files. If youâ€™ve configured
a remote Git repository (see <code>?wflow_git_remote</code>), click on
the hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/zq2209/omics-imputation-paper/blob/f5c067b0c080658e0a04d4c424336cdfb54eae4b/analysis/real_data_xqtl.Rmd" target="_blank">f5c067b</a>
</td>
<td>
â€œJohn
</td>
<td>
2025-04-13
</td>
<td>
updates
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>This page documents the procedures for xQTL analysis on real data,
including:</p>
<ul>
<li><p>Phenotype processing</p></li>
<li><p>Genotype PCA processing</p></li>
<li><p>Covariates processing</p></li>
<li><p>Cis association analysis with TensorQTL</p></li>
</ul>
<p>For detailed protocol, refer to <a
href="https://github.com/StatFunGen/xqtl-protocol">FunGen-xQTL
Computational Protocol</a>, which is molecular QTL analyses for the
NIH/NIA Alzheimerâ€™s Disease Sequencing Project (ADSP) Functional
Genomics xQTL (FunGen-xQTL) Project.</p>
<p>The pipelines in this protocol are written in the <a
href="https://vatlab.github.io/sos-docs/">Script of Scripts (SoS)</a>
workflow language. SoS can distribute and run jobs directly on
high-performance computing (HPC) clusters.</p>
<p>phenotype data for this analysis is <a
href="https://www.nature.com/articles/s41591-020-0815-6#Abs1">ROSMAP
Proteomics</a> from DLPFC region.</p>
<div id="phenotype-processing" class="section level3">
<h3>Phenotype Processing</h3>
<p>Data Input:</p>
<ul>
<li><p>Processed phenotype data from ROSMAP DLPFC subjects:
combine_r1andr2_protein.reg_cov_cog.uniq.proj.csv</p></li>
<li><p>Sample metadata: sampleSheetAfterQc.csv</p></li>
<li><p>Gene annotation file:
Homo_sapiens.GRCh38.103.chr.reformatted.collapse_only.gene.gtf</p></li>
</ul>
<p>First, we annotate each protein with chromosome, start, end positions
and gene ID.</p>
<pre class="bash"><code>sos run xqtl-protocol/pipeline/gene_annotation.ipynb annotate_coord_protein \
    --cwd proteomics/rosmap/pheno \
    --phenoFile combine_r1andr2_protein.reg_cov_cog.uniq.proj.csv \
    --annotation-gtf Homo_sapiens.GRCh38.103.chr.reformatted.collapse_only.gene.gtf \
    --container containers/rna_quantification.sif \
    --phenotype-id-type gene_name \
    --sample-participant-lookup sampleSheetAfterQC.txt</code></pre>
<p>Then, we impute the phenotype data by using each method. Refer to
Section
<code>Impute Dataset with Generated Missing by Using Different Methods</code>
in <a
href="https://zq2209.github.io/omics-imputation-paper/imputation_accuracy_benchmark.html">imputation
accuracy benchmark</a> for details on how to impute missing data. The
output of imputation is <code>rosmap_pheno_imp.bed.gz</code>.</p>
<p>We then split the imputed phenotype data by chromosome to be meet the
required format of the pipeline.</p>
<pre class="python"><code>for i, g in pheno.groupby(&#39;#chr&#39;):
    g.to_csv(f&#39;rosmap/pheno/rosmap_pheno_{i}.bed&#39;, index=None, sep=&#39;\t&#39;)</code></pre>
<pre class="bash"><code># Compress and index
bgzip rosmap/pheno/*.bed
tabix rosmap/pheno/*.bed.gz</code></pre>
</div>
<div id="genotype-pca-processing" class="section level3">
<h3>Genotype PCA Processing</h3>
<p>Data Input:</p>
<ul>
<li><p>Genotype files:</p>
<ul>
<li><p>ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.bed</p></li>
<li><p>ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.bim</p></li>
<li><p>ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.fam</p></li>
</ul></li>
</ul>
<p>Before getting PCs from genotype data, we match samples to identify
samples present in both genotype and phenotype data.</p>
<pre class="bash"><code>sos run xqtl-protocol/pipeline/GWAS_QC.ipynb genotype_phenotype_sample_overlap \
    --cwd /rosmap/geno \
    --phenoFile /rosmap/pheno/rosmap_pheno_imp.bed.gz \
    --genoFile ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.fam \
    --container containers/bioinfo.sif </code></pre>
<p>Then, we filter out related samples using <code>Kingship</code>.</p>
<pre class="bash"><code>sos run xqtl-protocol/pipeline/GWAS_QC.ipynb king \
    --cwd /rosmap/geno \
    --genoFile ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.bed \
    --name pQTL \
    --keep-samples /rosmap/geno/rosmap_pheno.sample_genotypes.txt \
    --container containers/bioinfo.sif --no-maximize-unrelated</code></pre>
<p>To further quality control genotype data, we filter variants with MAC
&lt; 5 and perform LD pruning.</p>
<pre class="bash"><code>sos run xqtl-protocol/pipeline/GWAS_QC.ipynb qc \
    --cwd rosmap/geno \
    --genoFile ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.bed \
    --keep-samples /rosmap/geno/rosmap_pheno.sample_genotypes.txt \
    --mac-filter 5 \
    --container containers/bioinfo.sif</code></pre>
<p>After all QC steps, we perform PCA on pruned genotype data to
generate genetic covariates.</p>
<pre class="bash"><code>sos run xqtl-protocol/pipeline/PCA.ipynb flashpca \
    --cwd /rosmap/pca \
    --genoFile ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.plink_qc.prune.bed \
    --container containers/flashpcaR.sif</code></pre>
</div>
<div id="covariates-preprocessing" class="section level3">
<h3>Covariates Preprocessing</h3>
<p>Data Input:</p>
<ul>
<li><p>Imputed phenotype data:
<code>rosmap_pheno_imp.bed.gz</code></p></li>
<li><p>Genotype PCs from <code>Genotype PCA Processing</code></p></li>
<li><p>Known covariates: sex, age at death, PMI in
<code>rosmap_cov.txt</code></p></li>
</ul>
<p>We first combine known covariates with genotype PCs.</p>
<pre class="bash"><code>sos run xqtl-pipeline/pipeline/covariate_formatting.ipynb merge_genotype_pc \
    --cwd ./rosmap/cov \
    --pcaFile ./rosmap/pca/ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.plink_qc.prune.pca.rds \
    --covFile ./rosmap/cov/rosmap_cov.txt \
    --tol_cov 0.4 \
    --k `awk &#39;$3 &lt; 0.7&#39; ./rosmap/pca/ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.plink_qc.prune.pca.scree.txt | tail -1 | cut -f 1` \
    --container ./bioinfo.sif</code></pre>
<p>Then, we identify hidden factors (i.e.Â batch effects) using PCA. The
number of hidden factors is determined by Marchenko-Pastur method. And
the residual after regressing out all covariates (known covariates,
genotype PCs, hidden factors) is calculated.</p>
<pre class="bash"><code>sos run xqtl-pipeline/pipeline/covariate_hidden_factor.ipynb Marchenko_PC \
    --cwd ./rosmap/cov \
    --phenoFile ./rosmap/pheno/rosmap_pheno_imp.bed.gz \
    --container ./PCAtools.sif \
    --covFile ./rosmap/cov/rosmap_cov.ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.plink_qc.prune.pca.gz</code></pre>
</div>
<div id="cis-association-analysis-with-tensorqtl"
class="section level3">
<h3>Cis Association Analysis with TensorQTL</h3>
<p>Data Input: - Genotype:
<code>ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.plink_files_list.txt</code></p>
<ul>
<li><p>Phenotype: <code>rosmap/rosmap_pheno_imp.txt</code></p></li>
<li><p>Covariates: Output from
<code>Covariates Processing</code></p></li>
<li><p>Custom cis-window (TAD Window):
<code>TAD_TADB/extended_cis_before_winsorize.tsv</code></p></li>
</ul>
<pre class="bash"><code>sos run xqtl-protocol/pipeline/TensorQTL.ipynb cis \
    --genotype-file ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.plink_files_list.txt \
    --phenotype-file rosmap/rosmap_pheno_imp.txt \
    --covariate-file rosmap/cov/rosmap_pheno.rosmap_cov.ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.pQTL.unrelated.filtered.prune.pca.resid.Marchenko_pc.gz \
    --customized_cis_list TAD_TADB/extended_cis_before_winsorize.tsv \
    --cwd pQTL_cis/TADB \
    --container containers/TensorQTL.sif --MAC 5</code></pre>
<p>Finally, we extract the number of QTL detected.</p>
<pre class="bash"><code>cat pheno_recipe_rosmap_pheno.rosmap_cov.ROSMAP_NIA_WGS.leftnorm.filtered.pQTL.unrelated.filtered.prune.pca.resid.Marchenko_pc.1.n_sig.txt</code></pre>
<p>This whole procedure is applied to all 8 imputed ROSMAP phenotype
data by using 8 imputation methods. The numbers of QTL significant genes
are used to compare performance of each method.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.3 (2025-02-28)
Platform: aarch64-apple-darwin20
Running under: macOS Sequoia 15.4

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/Chicago
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] vctrs_0.6.5       cli_3.6.4         knitr_1.50        rlang_1.1.5      
 [5] xfun_0.51         stringi_1.8.7     png_0.1-8         promises_1.3.2   
 [9] jsonlite_2.0.0    workflowr_1.7.1   glue_1.8.0        rprojroot_2.0.4  
[13] git2r_0.36.2      htmltools_0.5.8.1 httpuv_1.6.15     sass_0.4.9       
[17] rmarkdown_2.29    grid_4.4.3        evaluate_1.0.3    jquerylib_0.1.4  
[21] tibble_3.2.1      fastmap_1.2.0     yaml_2.3.10       lifecycle_1.0.4  
[25] whisker_0.4.1     stringr_1.5.1     compiler_4.4.3    fs_1.6.5         
[29] Rcpp_1.0.14       pkgconfig_2.0.3   rstudioapi_0.17.1 later_1.4.1      
[33] lattice_0.22-6    digest_0.6.37     R6_2.6.1          reticulate_1.42.0
[37] pillar_1.10.1     magrittr_2.0.3    Matrix_1.7-3      bslib_0.9.0      
[41] tools_4.4.3       cachem_1.1.0     </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
