---
title: "Replication analysis for xQTL"
author: "Zining Qi"
output: workflowr::wflow_html
---

In the context of real-world xQTL discoveries, the absence of known ground truth poses a challenge in validating the robustness of different imputation methods for the xQTL signals obtained through them. To address this, we conduct replication analyses on proteomics data from ROSMAP. This page documents the procedure of replication analysis. 

### Setup and Initial Data Loading

First, load the packages needed for this analysis and the original phenotype data.

```{r, eval=FALSE}
library(tidyverse)
library(data.table)
library(qvalue)
library(stringr)
library(patchwork)

pheno <- read_delim('combine_r1andr2_protein.reg_cov_cog.uniq.proj.bed.gz')
```

### Replication Analysis on Strategy A

##### Define Standard Set

We defined the standard set as the pQTL result using subset of observed samples in proteomics data. We focus on the proteins with $>350$ observed samples. For proteins with more than 350 observed samples, QTL analysis is conducted on observed samples.

First, we split the phenotype by each features (i.e. proteins) with more than 350 observed samples.

```{r, eval=FALSE}
write_bed <- function(data_row, file_name) {
  # Write the row to a BED file
  write_delim(data_row, file_name, delim = '\t')
}
```

```{r, eval=FALSE}
gene_list <- c()
id_list <- c()
chr_list <- c()
for (i in 1:nrow(pheno)) {
  data_row <- pheno[i, ]

  # Calculate number of exsiting samples
  num <- sum(!is.na(data_row[, -c(1:4)]))

  # Create file name
  file_name <- paste0('strategy1/input_per_gene/standard/rosmap_pheno.', data_row$ID, '.bed')
  ensg_id <- str_extract(data_row$ID, "ENSG[0-9]+")

  # Write the row to a BED file if there are more than 350 exsiting samples
  if (length(data_row) > 350) {
    write_bed(data_row, file_name)
    gene_list <- c(gene_list, ensg_id)
    id_list <- c(id_list, data_row$ID)
    chr_list <- c(chr_list, data_row$chr)
  }
}
```

Then, we run cis-QTL analysis on each protein with more than 350 samples by using [FunGen-xQTL Computational Protocol](https://github.com/StatFunGen/xqtl-protocol). 

```{r, eval=FALSE}
for (i in length(gene_lst)) {
    sos_command <- paste0(
        "sos run xqtl-protocol/pipeline/TensorQTL.ipynb cis:1  ",
        "--genotype-file ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.", gene_list[i], ".bed ",
        "--phenotype-file strategy1/input_per_gene/standard/rosmap_pheno.", id_list[i], ".bed.gz ",
        "--covariate-file cov/soft/rosmap_soft.resid.Marchenko_PC.gz ",
        "--customized_cis_windows ROSMAP_pQTL_TADB_enhanced_cis.bed ",
        "--cwd strategy1/output/standard ",
        "--container containers/tensorqtl.sif --MAC 5 --chromosome ", chr_list[i],
        " -s build", 
        sep = " "
      )

    sbatch_script <- paste(
        "#!/bin/bash\n",
        "#SBATCH --job-name=tensor_", gene_list[i], "\n",
        "#SBATCH --mem=30G\n",
        "#SBATCH --time=10:00:00\n",
        "#SBATCH --output=output/tensor_%j.out\n",
        "#SBATCH --error=output/tensor_%j.err\n",
        "#SBATCH -p CSG\n\n",
        "export PATH=
PATH\n",
        "source ~/.bashrc\n",
        "module load Singularity/3.9.4\n",
        "cd ~\n\n",
        sos_command,
        sep = ""
    )
    
    writeLines(sbatch_script, con = file.path('strategy1/script/', paste0("tensor_", id_list[i], ".sbatch")))
}
```

The result from this step is considered as replication set. 

##### Define Imputed Set

For each of 8 imputation methods, we imputed the proteomics data first by using each method. Refer to Section `Impute Dataset with Generated Missing by Using Different Methods` in [imputation accuracy benchmark](https://zq2209.github.io/omics-imputation-paper/imputation_accuracy_benchmark.html) for details on how to impute missing data. The output of imputation is `rosmap_pheno_{method}_imp.bed.gz`.

And then we randomly selected subsets of samples with different sample sizes (eg, N=200, 250, 300) for proteins to conduct QTL analysis. 

```{r, eval=FALSE}
n <- 200, 250, 300
for (mtd in c("ebmf", "mofa", "xgb", "rf", "soft", "knn", "mean", "lod")) {
  pheno <- fread(paste0('rosmap_pheno_', mtd, '_imp.bed.gz'))
  samples <- colnames(pheno)[-c(1:4)]
  
  for (i in 1:nrow(pheno)) {
    data_row <- pheno[i, ]
  
    # randomly select samples
    sel_samples <- c(colnames(pheno)[c(1:4)], sample(samples, n))
    data_row <- pheno[i, sel_samples]
    # Create file name
    file_name <- paste0('strategy1/input_per_gene/', mtd, '/rosmap_pheno.', data_row$ID, '.bed')
  
    # Write the row to a BED file if there are more than 350 exsiting samples
    if (data_row$ID %in% chr_list) {
      write_bed(data_row, file_name)
    }
  }
}
```

Before running cis-QTL analysis for each method, we identify hidden factors and calculate residuals first.  

```{r, eval=FALSE}
for (mtd in c("ebmf", "mofa", "xgb", "rf", "soft", "knn", "mean", "lod")) {
    hidden_sos <- paste0(
        "sos run xqtl-protocol/pipeline/covariate_hidden_factor.ipynb Marchenko_PC  ",
        "--phenoFile strategy1/rosmap_pheno.", mtd, "_imp.bed.gz ",
        "--covFile cov/ROSMAP_cov.ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.rosmap_proteomics.unrelated.plink_qc.prune.pca.gz ",
        "--cwd strategy1/input_", mtd, " ",
        "--container containers/pcatools.sif -J 50",
        sep = " "
    )
    
    sbatch_script <- paste(
        "#!/bin/bash\n",
        "#SBATCH --job-name=hidden_", mtd, "\n",
        "#SBATCH --mem=30G\n",
        "#SBATCH --time=10:00:00\n",
        "#SBATCH --output=/hidden_%j.out\n",
        "#SBATCH --error=/hidden_%j.err\n",
        "#SBATCH -p CSG\n\n",
        "export PATH=
PATH\n",
        "source ~/.bashrc\n",
        "module load Singularity/3.9.4\n",
        "cd ~\n\n",
        hidden_sos,
        sep = ""
    )
    
    writeLines(sbatch_script, con = file.path('strategy1/script_imp/', paste0("hidden_", mtd, ".sbatch")))
}
```


Next step is to run cis-QTL analysis for each method. 

```{r, eval=FALSE}
for (mtd in c("ebmf", "mofa", "xgb", "rf", "soft", "knn", "mean", "lod")) {
    for (i in length(gene_lst)) {
        sos <- paste0(
            "sos run xqtl-protocol/pipeline/TensorQTL.ipynb cis:1  ",
            "--genotype-file ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.", gene_list[i], ".bed ",
            "--phenotype-file strategy1/input_per_gene/", mtd, '/rosmap_pheno.', data_row$ID, ".bed.gz ",
            "--covariate-file strategy1/input_", mtd, "/rosmap_pheno.ROSMAP_cov.ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.rosmap_proteomics.unrelated.plink_qc.prune.pca.Marchenko_PC.gz ",
            "--customized_cis_windows ROSMAP_pQTL_TADB_enhanced_cis.bed ",
            "--cwd strategy1/", mtd, " ",
            "--container containers/tensorqtl.sif --MAC 5 --chromosome ", i,
            " -s build", 
            sep = " "
          )

        sbatch_script <- paste(
            "#!/bin/bash\n",
            "#SBATCH --job-name=tensor_", mtd, i, "\n",
            "#SBATCH --mem=80G\n",
            "#SBATCH --time=10:00:00\n",
            "#SBATCH --output=tensor_%j.out\n",
            "#SBATCH --error=tensor_%j.err\n",
            "#SBATCH -p GEN\n\n",
            "export PATH=
PATH\n",
            "source ~/.bashrc\n",
            "module load Singularity/3.9.4\n",
            "cd ~\n\n",
            sos,
            sep = ""
        )

        writeLines(sbatch_script, con = file.path('strategy1/script_imp/', paste0(mtd, "_tensor_", i, ".sbatch")))
    }
}
```

This step gives us 8 discovery sets for each sample size. 

##### Calculate the replicate rate

For each of the discovery sets, we identify significant pairs as FDR $< 0.05$, which gives us 8 sets of significant pairs. Then, we calculate the proportion of significant pairs that are still significant (FDR $< 0.05$) in the replication set. The formula for the replication rate is:

$$ \text{Replication rate} = \frac{\#\text{significant pairs in both discovery and replication sets}}{\#\text{significant pairs in the discovery set}}$$

To calculate this replication rate, we first identify significant pairs in standard set. 

```{r, eval=FALSE}
standard.sum <- list.files(path = 'strategy1/output/standard', pattern = "rosmap_pheno.*.cis_qtl.regional.tsv.gz$", full.names = TRUE)

standard.lst <- lapply(standard.sum, fread)
standard <- rbindlist(standard.lst)

standard$id <- paste0(standard$molecular_trait_id, '-', standard$variant_id)
standard <- standard %>% group_by(molecular_trait_id) %>% mutate(fdr = p.adjust(pvalue, method = "fdr"))
standard.sig <- standard %>% filter(fdr < 0.05)
```

Finally, we calculate singificant pairs in discovery sets and check how many significant pairs in discoveries sets are still significant in replication set. 

```{r, eval=FALSE}
replication_rates <- c()
for (mtd in c("ebmf", "mofa", "xgb", "rf", "soft", "knn", "mean", "lod")) {
  imp.sum <- list.files(path = paste0("strategy1/", mtd), pattern = paste0("rosmap_", mtd, ".*.cis_qtl.regional.tsv.gz$"), full.names = TRUE)
  imp.lst <- lapply(imp.sum, fread)
  imp <- rbindlist(imp.lst)
  
  imp <- imp %>% filter(molecular_trait_object_id %in% standard$molecular_trait_object_id)
  imp <- imp %>% group_by(molecular_trait_id) %>% mutate(fdr = p.adjust(pvalue, method = "fdr"))
  imp$id <- paste0(imp$molecular_trait_id, '-', imp$variant_id)
  imp.sig <- imp %>% filter(fdr < 0.05)
  
  pairs.sig <- unique(imp.sig$id)
  replicated.pairs <- noimp %>% filter(id %in% pairs.sig)
  rate <- mean(replicated.pairs$fdr < 0.05)
  replication_rates <- c(replication_rates, rate)
}
```


### Replication Analysis on Strategy B

In addition this strategy, we designed another way to evaluate replication for each of 8 methods.

##### Extract Phenotypes that Have No Missing Samples

```{r, eval=FALSE}
pheno.comp <- pheno[complete.cases(pheno), ]
```


##### Define Standard Set

The standard set is defined same as Strategy A. 

```{r, eval=FALSE}
gene_list <- c()
id_list <- c()
chr_list <- c()
for (i in 1:nrow(pheno)) {
  data_row <- pheno[i, ]

  # Calculate number of exsiting samples
  num <- sum(!is.na(data_row[, -c(1:4)]))

  # Create file name
  file_name <- paste0('strategy2/input_per_gene/standard/rosmap_pheno.', data_row$ID, '.bed')
  ensg_id <- str_extract(data_row$ID, "ENSG[0-9]+")

  # Write the row to a BED file if there are more than 350 exsiting samples
  if (length(data_row) > 350) {
    write_bed(data_row, file_name)
    gene_list <- c(gene_list, ensg_id)
    id_list <- c(id_list, data_row$ID)
    chr_list <- c(chr_list, data_row$chr)
  }
}
```

Then, we run cis-QTL analysis on each protein with all exsting samples by using [FunGen-xQTL Computational Protocol](https://github.com/StatFunGen/xqtl-protocol). 

```{r, eval=FALSE}
for (i in length(gene_lst)) {
    sos_command <- paste0(
        "sos run xqtl-protocol/pipeline/TensorQTL.ipynb cis:1  ",
        "--genotype-file ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.", gene_list[i], ".bed ",
        "--phenotype-file strategy2/input_per_gene/standard/rosmap_pheno.", id_list[i], ".bed.gz ",
        "--covariate-file cov/soft/rosmap_soft.resid.Marchenko_PC.gz ",
        "--customized_cis_windows ROSMAP_pQTL_TADB_enhanced_cis.bed ",
        "--cwd strategy2/output/standard ",
        "--container containers/tensorqtl.sif --MAC 5 --chromosome ", chr_list[i],
        " -s build", 
        sep = " "
      )

    sbatch_script <- paste(
        "#!/bin/bash\n",
        "#SBATCH --job-name=tensor_", gene_list[i], "\n",
        "#SBATCH --mem=30G\n",
        "#SBATCH --time=10:00:00\n",
        "#SBATCH --output=output/tensor_%j.out\n",
        "#SBATCH --error=output/tensor_%j.err\n",
        "#SBATCH -p CSG\n\n",
        "export PATH=
PATH\n",
        "source ~/.bashrc\n",
        "module load Singularity/3.9.4\n",
        "cd ~\n\n",
        sos_command,
        sep = ""
    )
    
    writeLines(sbatch_script, con = file.path('strategy2/script/', paste0("tensor_", id_list[i], ".sbatch")))
}
```

The result from this step is considered as replication set. 


##### Define Imputed Set

We randomly assign N = 50, 100, 150 missing samples for each protein. 

```{r, eval=FALSE}
n_miss <- 50, 100, 150

for (i in nrow(pheno.comp)) {
    # Randomly select missing
    missing_samples <- sample(colnames(pheno.comp)[-c(1:4)], n_miss)
    # Set these to NA
    pheno_miss[i, missing_samples] <- NA
}
```

We then imputed the proteomics data first by using each method. Refer to Section `Impute Dataset with Generated Missing by Using Different Methods` in [imputation accuracy benchmark](https://zq2209.github.io/omics-imputation-paper/imputation_accuracy_benchmark.html) for details on how to impute missing data. The output of imputation is `rosmap_pheno_{method}_imp.bed.gz`.

```{r, eval=FALSE}
n_miss <- 50, 100, 150
n <- 350
sample_lst <- colnames(pheno.miss[, 5:ncol(pheno.miss)])
sample.all <- list()
for (i in 1:nrow(pheno.miss)) {
    row <- pheno[i, 5:ncol(pheno)]
    #id <- pheno[i, 1:4]
    # randomly select non-missing samples
    non_missing_samples <- sample_lst[!as.vector(is.na(row))]   
    non_missing <- sample(non_missing_samples, size = n - n_miss, replace = FALSE)
    # missing samples
    missing <- sample_lst[as.vector(is.na(row))]
    samples <- c(non_missing, missing)
    sample.all[[i]] <- samples
}
```

The imputed phenotype for QTL analysis includes number of imputed samples (i.e. n_miss = 50, 100, 150) plus randomly selected 350-n_miss samples from the existing samples.

```{r}
for (mtd in c("ebmf", "mofa", "xgb", "rf", "soft", "knn", "mean", "lod")) {
    pheno.imp <- read_delim(paste0('strategy2/rosmap_pheno_', mtd, '_imp.bed.gz'))
    
    for (i in 1:nrow(pheno.imp)) {
    # Get the list of samples for the current row
        samples <- sample.all[[i]]
        # Select the values from the dataframe for the current row based on sample names
        id <- pheno.imp[i, 1:4]
        selected_values <- pheno.imp[i, samples, drop = FALSE]
        data_df <- cbind(id, selected_values)
        # Save the new row dataframe as a BED file
        file_name <- paste0('strategy2/', mtd, '/rosmap_', mtd, '.', data_df$ID, '.bed')
        write_bed(data_df, file_name)
    }
}
```


Before running cis-QTL analysis for each method, we identify hidden factors and calculate residuals first.  

```{r, eval=FALSE}
for (mtd in c("ebmf", "mofa", "xgb", "rf", "soft", "knn", "mean", "lod")) {
    hidden_sos <- paste0(
        "sos run xqtl-protocol/pipeline/covariate_hidden_factor.ipynb Marchenko_PC  ",
        "--phenoFile strategy2/rosmap_pheno.", mtd, "_imp.bed.gz ",
        "--covFile cov/ROSMAP_cov.ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.rosmap_proteomics.unrelated.plink_qc.prune.pca.gz ",
        "--cwd strategy2/input_", mtd, " ",
        "--container containers/pcatools.sif -J 50",
        sep = " "
    )
    
    sbatch_script <- paste(
        "#!/bin/bash\n",
        "#SBATCH --job-name=hidden_", mtd, "\n",
        "#SBATCH --mem=30G\n",
        "#SBATCH --time=10:00:00\n",
        "#SBATCH --output=/hidden_%j.out\n",
        "#SBATCH --error=/hidden_%j.err\n",
        "#SBATCH -p CSG\n\n",
        "export PATH=
PATH\n",
        "source ~/.bashrc\n",
        "module load Singularity/3.9.4\n",
        "cd ~\n\n",
        hidden_sos,
        sep = ""
    )
    
    writeLines(sbatch_script, con = file.path('strategy2/script_imp/', paste0("hidden_", mtd, ".sbatch")))
}
```


Next step is to run cis-QTL analysis for each method. 

```{r, eval=FALSE}
for (mtd in c("ebmf", "mofa", "xgb", "rf", "soft", "knn", "mean", "lod")) {
    for (i in length(gene_lst)) {
        sos <- paste0(
            "sos run xqtl-protocol/pipeline/TensorQTL.ipynb cis:1  ",
            "--genotype-file ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.", gene_list[i], ".bed ",
            "--phenotype-file strategy2/input_per_gene/", mtd, '/rosmap_pheno.', data_row$ID, ".bed.gz ",
            "--covariate-file strategy2/input_", mtd, "/rosmap_pheno.ROSMAP_cov.ROSMAP_NIA_WGS.leftnorm.bcftools_qc.plink_qc.rosmap_proteomics.unrelated.plink_qc.prune.pca.Marchenko_PC.gz ",
            "--customized_cis_windows ROSMAP_pQTL_TADB_enhanced_cis.bed ",
            "--cwd strategy2/", mtd, " ",
            "--container containers/tensorqtl.sif --MAC 5 --chromosome ", i,
            " -s build", 
            sep = " "
          )

        sbatch_script <- paste(
            "#!/bin/bash\n",
            "#SBATCH --job-name=tensor_", mtd, i, "\n",
            "#SBATCH --mem=80G\n",
            "#SBATCH --time=10:00:00\n",
            "#SBATCH --output=tensor_%j.out\n",
            "#SBATCH --error=tensor_%j.err\n",
            "#SBATCH -p GEN\n\n",
            "export PATH=
PATH\n",
            "source ~/.bashrc\n",
            "module load Singularity/3.9.4\n",
            "cd ~\n\n",
            sos,
            sep = ""
        )

        writeLines(sbatch_script, con = file.path('strategy1/script_imp/', paste0(mtd, "_tensor_", i, ".sbatch")))
    }
}
```

This step gives us 8 discovery sets for each number of missing assignment. 


##### Calculate the replicate rate

Similar to the previous strategy, we calculate the replication rate by using the same formula.

To calculate this replication rate, we first identify significant pairs in standard set. 

```{r, eval=FALSE}
standard.sum <- list.files(path = 'strategy2/output/standard', pattern = "rosmap_pheno.*.cis_qtl.regional.tsv.gz$", full.names = TRUE)

standard.lst <- lapply(standard.sum, fread)
standard <- rbindlist(standard.lst)

standard$id <- paste0(standard$molecular_trait_id, '-', standard$variant_id)
standard <- standard %>% group_by(molecular_trait_id) %>% mutate(fdr = p.adjust(pvalue, method = "fdr"))
standard.sig <- standard %>% filter(fdr < 0.05)
```

Finally, we calculate singificant pairs in discovery sets and check how many significant pairs in discoveries sets are still significant in replication set. 

```{r, eval=FALSE}
replication_rates <- c()
for (mtd in c("ebmf", "mofa", "xgb", "rf", "soft", "knn", "mean", "lod")) {
  imp.sum <- list.files(path = paste0("strategy2/", mtd), pattern = paste0("rosmap_", mtd, ".*.cis_qtl.regional.tsv.gz$"), full.names = TRUE)
  imp.lst <- lapply(imp.sum, fread)
  imp <- rbindlist(imp.lst)
  
  imp <- imp %>% filter(molecular_trait_object_id %in% standard$molecular_trait_object_id)
  imp <- imp %>% group_by(molecular_trait_id) %>% mutate(fdr = p.adjust(pvalue, method = "fdr"))
  imp$id <- paste0(imp$molecular_trait_id, '-', imp$variant_id)
  imp.sig <- imp %>% filter(fdr < 0.05)
  
  pairs.sig <- unique(imp.sig$id)
  replicated.pairs <- noimp %>% filter(id %in% pairs.sig)
  rate <- mean(replicated.pairs$fdr < 0.05)
  replication_rates <- c(replication_rates, rate)
}
```



### Visualization of Replication Results

```{r, eval=FALSE}
# Define the methods and values
methods <- c('gEBMF', 'MOFA2', "SoftImpute", "MissXGB", 'MissForest', "KNN", "MeanImpute", 'LOD')

data1 <- data.frame(
  Method = factor(rep(methods, times = 3), levels = methods),
  Value = c(replication_rates_s1_200, replication_rates_s1_250, replication_rates_s1_300),
  Sample = factor(rep(c("200", "250", "300"), each = length(methods)))
)


data2 <- data.frame(
  Method = factor(rep(methods, times = 3), levels = methods),
  Value = c(replication_rates_s1_50, replication_rates_s1_100, replication_rates_s1_150),
  Missing = factor(rep(c("50", "100", "150"), each = length(methods)), levels = c('50', '100', '150'))
)

rep_1 <- ggplot(data1, aes(Method, Value, fill = Sample)) +
  geom_bar(stat="identity", position = "dodge") + 
  labs(title = "Replication",
       x = "Method",
       y = "Replication Rate") +
  theme_minimal() +
  coord_cartesian(ylim =c (0.6, 1)) +
  theme(axis.text.x = element_text(angle = 60, hjust = 1))

rep_2 <- ggplot(data2, aes(Method, Value, fill = Missing)) +
  geom_bar(stat="identity", position = "dodge") + 
  labs(title = "Replication",
       x = "Method",
       y = "Replication Rate") +
  theme_minimal() +
  coord_cartesian(ylim =c (0.6, 1)) +
  theme(axis.text.x = element_text(angle = 60, hjust = 1))

combined_plot <- (rep_1 | rep_2) 
print(combined_plot)
```


### Example Output

![](figure/replication.png){width="60%"}




